### PR TITLE
Start abstracting platform logic builds behind a shared interface (II)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

To move to a world where platform builds are pluggable, we'll need to start defining them in terms of a unified interface. Here, I've added the capability to select a step based on the target platform and build mode, this can almost be done without knowledge of the device.

To follow this up, we'll need to determine what device specific and build specific results are interesting enough. For example, for android we need to know something like the launchActivity. For other platforms we might only need the executable.

## Related Issues

#25377

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Update: use device targetplatform if one is not specified. bundle/apk already default to android_arm

File: packages/flutter_tools/lib/src/android/apk.dart [hunk 1]
```diff
@@ -4,38 +4,74 @@
...
 
-Future<void> buildApk({
-  @required FlutterProject project,
-  @required String target,
-  BuildInfo buildInfo = BuildInfo.debug,
-}) async {
-  if (!project.android.isUsingGradle) {
-    throwToolExit(
+/// The Android Gradle build step.
+class AndroidPlatformBuildStep extends PlatformBuildStep {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/apk.dart]: Maybe PlatformBuildStep -> BuildStep ?

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 2]
```diff
@@ -358,26 +358,25 @@ class AndroidDevice extends Device {
...
       printTrace('Building APK');
       final FlutterProject project = FlutterProject.current();
-      await buildApk(
-          project: project,
-          target: mainPath,
-          buildInfo: buildInfo,
+      PlatformBuildStep platformStep;
+      try {
+        platformStep = platformBuilders.selectPlatform(buildInfo: buildInfo);
+      } on Exception catch (err) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/android/android_device.dart]: Catch the specific exceptions, or change `selectPlatform` to return an object that distinguishes among the different kinds of errors.

File: packages/flutter_tools/lib/src/commands/build_apk.dart [hunk 3]
```diff
@@ -47,10 +49,17 @@ class BuildApkCommand extends BuildSubCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
-    await buildApk(
+    final BuildInfo buildInfo = getBuildInfo();
+    PlatformBuildStep platformStep;
+    try {
+      platformStep = platformBuilders.selectPlatform(buildInfo: buildInfo);
+    } on Exception catch (err) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/build_apk.dart]: ditto

File: packages/flutter_tools/lib/src/platform_step.dart [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        buildInfo.targetPlatform,
+      );
+    }
+    return selected;
+  }
+}
+
+/// An exception thrown when there is no builder corresponding to the provided
+/// target platform and build mode combination.
+class MissingPlatformBuilder implements Exception {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/platform_step.dart]: Maybe subclass, like "PlatformBuilderException" to make it easier to catch just the right things.

File: packages/flutter_tools/lib/src/platform_step.dart [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import 'build_info.dart';
+import 'cache.dart';
+import 'project.dart';
+
+/// The [PlatformBuilders] instance.
+PlatformBuilders get platformBuilders => context.get<PlatformBuilders>();
+
+/// A registry that selects the correct [PlatformBuilder] based on the provided
+/// build info.
+class PlatformBuilders {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/platform_step.dart]: Maybe just "Builders"?

File: packages/flutter_tools/lib/src/platform_step.dart [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  /// Create a new [PlatformBuilder] with `_platformSteps` as platform specific
+  /// implementations.
+  const PlatformBuilders([this._platformSteps = const <PlatformBuildStep>[
+    AndroidPlatformBuildStep(),
+  ]]);
+
+  final List<PlatformBuildStep> _platformSteps;
+
+  /// Build the platform specific bundle for [flutterProject].
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/platform_step.dart]: This comment doesn't seem accurate.

File: packages/flutter_tools/lib/src/platform_step.dart [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  ///
+  /// Selects the [PlatformBuildStep] that supports the required [TargetPlatform]
+  /// and [BuildMode] requested.
+  ///
+  /// Throws an [AmbiguousPlatformBuilder] if multiple build steps
+  /// match.
+  ///
+  /// Throws a [MissingPlatformBuilder] if no build steps match.
+  PlatformBuildStep selectPlatform({
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/platform_step.dart]: How about "selectBuildStep"

Thinking ahead, this method may in the future also need to understand a concept of the stage of the build. E.g. something -> source, source -> kernel, kernel -> kernel, kernel -> native, kernel -> something platform specific.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/platform_step.dart]: This sounds like the right direction, but our current implementation of most native builds is somewhat backwards. For example, on Android we invoke gradle which re-entrantly invokes flutter build apk.

In this structure, the platform build comes first, and contains all of the other build steps, then returns to the platform step.

**Comment by jonahwilliams** [comment]
Closing as this is not quite the right approach

**Pull request closed without merging**